### Q1.a)  Use this function to find minima for (i) x2 + 3x+4 and (ii) x4 – 3x2 +2x. [5 points]


In [23]:
import numpy as np
import time
import random



In [24]:


gradient1 = lambda x: 2*x+3 ## specifies gradient for the function x2 + 3x+4
gradient2 = lambda x: 4*(x**3)-6*x+2 ## specifies gradient for the function  x4 – 3x2 +2x.

def gradient_descent (gradient, init_, learn_rate, n_iter=1000, tol=1e-06):
    x = init_
    for i in range (n_iter): 
        delta = (-1)*learn_rate * gradient(x)
        if np.all(np.abs (delta) <= tol): 
            break 
        x += delta 
    print('Iterations:', str(i+1))
    return  round(x*1000)/1000


print('Minima for the function x^2+3x+4 is found at '+ str(gradient_descent(gradient1,4 , 0.1)))
print('\n')
print('Minima for the function x^4-3x^2+2x is found at '+ str(gradient_descent(gradient2,4 , 0.01)))




Iterations: 64
Minima for the function x^2+3x+4 is found at -1.5


Iterations: 158
Minima for the function x^4-3x^2+2x is found at 1.0


### Q1.b) Write a gradient function to calculate gradients for a linear regression y = ax + b [10 points]


In [25]:
def linear_regression_gradients_calculator(X, Y, learning_rate = 0.01, init_a = 0, init_b = 0, max_iter = 30, tol = 1e-06):
    a = 0
    b = 0

    n = float(len(X)) 
    
    for i in range(max_iter): 
        Y_pred = a*X + b 
        
        D_a = (-2/n) * sum(X * (Y - Y_pred))  # gradient wrt a
        D_b = (-2/n) * sum(Y - Y_pred)  # gradient wrt b
        
        print('Gradient wrt "a": ', str(D_a), 'Gradient wrt "b": ', str(D_b))
        
        delta_a = (-1)*learning_rate * D_a
        delta_b = (-1)*learning_rate * D_b

        if (np.all(np.abs (delta_a) <= tol) and np.all(np.abs (delta_b) <= tol)): 
            break
        
        a+=delta_a
        b+=delta_b
        
        

In [26]:
np. random.seed (0)
X = 2.5*np.random.randn(10000)+1.5 #Arra
res=1.5*np.random.randn(10000) # Gene.
Y = 2+0.3*X + res

In [27]:
linear_regression_gradients_calculator(X, Y)



Gradient wrt "a":  -10.732095840897884 Gradient wrt "b":  -4.905313437381427
Gradient wrt "a":  -8.827398630157248 Gradient wrt "b":  -4.495135915979009
Gradient wrt "a":  -7.247353142900043 Gradient wrt "b":  -4.148547328558394
Gradient wrt "a":  -5.936807080137242 Gradient wrt "b":  -3.85483557168645
Gradient wrt "a":  -4.849974863413731 Gradient wrt "b":  -3.6051065198660663
Gradient wrt "a":  -3.948846887502027 Gradient wrt "b":  -3.3919752975366517
Gradient wrt "a":  -3.201868928685552 Gradient wrt "b":  -3.209309981884463
Gradient wrt "a":  -2.582845828290958 Gradient wrt "b":  -3.052018832176958
Gradient wrt "a":  -2.070031362953744 Gradient wrt "b":  -2.9158736535452054
Gradient wrt "a":  -1.6453726816498069 Gradient wrt "b":  -2.7973631585288707
Gradient wrt "a":  -1.2938830595227675 Gradient wrt "b":  -2.6935712318872813
Gradient wrt "a":  -1.003121176551481 Gradient wrt "b":  -2.6020758693756925
Gradient wrt "a":  -0.7627588300156192 Gradient wrt "b":  -2.5208652794442905
Gr

### Q1.c) Generate artificial data for this regression according to the following protocol


In [29]:
def lr_gd(X, Y, learning_rate = 0.05, init_a = 0, init_b = 0, max_iter = 1000, tol = 1e-06):
    start_time = time.time()

    a = 0
    b = 0

    n = float(len(X)) 
    
    for i in range(max_iter): 
        Y_pred = a*X + b 
        
        D_a = (-2/n) * sum(X * (Y - Y_pred))  
        D_b = (-2/n) * sum(Y - Y_pred)  
        
        delta_a = (-1)*learning_rate * D_a
        delta_b = (-1)*learning_rate * D_b

        if (np.all(np.abs (delta_a) <= tol) and np.all(np.abs (delta_b) <= tol)): 
            break
        
        a+=delta_a
        b+=delta_b
# =============================================================================
#         print(i)
# =============================================================================
    print('It took ', str(i+1), ' iterations to arrive at the desired result')
    batch_time = time.time() - start_time
    print("--- %s seconds ---" % (time.time() - start_time))

    return (round(a,2), round(b,2))


In [30]:
 print('\nUsing batch gradient descent with a learning rate of 0.01, the values of (a,b) are ', str(lr_gd(X, Y)))

It took  160  iterations to arrive at the desired result
--- 0.23725533485412598 seconds ---

Using batch gradient descent with a learning rate of 0.01, the values of (a,b) are  (0.3, 2.02)


### Q1.d) Implement minibatch stochastic gradient descent using the code base you have developed so far.[15 points]


In [31]:
def lr_mb_gd(X, Y, batch_size, learning_rate = 0.05, init_a = 0, init_b = 0, max_iter = 1000, tol = 1e-06):
    start_time = time.time()
    a = 0
    b = 0

    n = float(len(X)) 
    
    for i in range(max_iter): 
        random_index=random.randint(0,n-batch_size-1)
        X_mini= X[random_index:random_index+batch_size]
        Y_mini = Y[random_index:random_index+batch_size]
        
        
        Y_pred = a*X_mini + b 

        D_a = (-2/n) * sum(X_mini * (Y_mini - Y_pred))  
        D_b = (-2/n) * sum(Y_mini - Y_pred)  

        delta_a = (-1)*learning_rate * D_a
        delta_b = (-1)*learning_rate * D_b

        if (np.all(np.abs (delta_a) <= tol) and np.all(np.abs (delta_b) <= tol)): 
            break

        a+=delta_a
        b+=delta_b
# =============================================================================
#         print(i)
# =============================================================================
    print('It took ', str(i+1), ' iterations to arrive at the desired result')
    print("--- %s seconds ---" % (time.time() - start_time))
    print (round(a,2), round(b,2))



## I will try various batch sizes(with a fixed learning rate __0.05__) and will finally pick a size that gives us the best values of _a_ and _b_

In [32]:
batch_sizes = [i for i in range(1,1001)]

In [33]:
for b in batch_sizes:
    print('batch_size ', str(b))
    lr_mb_gd(X, Y, b)
    print('\n')
    


batch_size  1
It took  165  iterations to arrive at the desired result
--- 0.003868579864501953 seconds ---
0.01 0.0


batch_size  2
It took  1000  iterations to arrive at the desired result
--- 0.01896953582763672 seconds ---
0.1 0.05


batch_size  3
It took  1000  iterations to arrive at the desired result
--- 0.02090930938720703 seconds ---
0.14 0.07


batch_size  4
It took  1000  iterations to arrive at the desired result
--- 0.01994633674621582 seconds ---
0.18 0.09


batch_size  5
It took  1000  iterations to arrive at the desired result
--- 0.024341106414794922 seconds ---
0.21 0.11


batch_size  6
It took  1000  iterations to arrive at the desired result
--- 0.01991748809814453 seconds ---
0.25 0.13


batch_size  7
It took  1000  iterations to arrive at the desired result
--- 0.019580841064453125 seconds ---
0.29 0.15


batch_size  8
It took  1000  iterations to arrive at the desired result
--- 0.024103879928588867 seconds ---
0.3 0.17


batch_size  9
It took  1000  iterations 

batch_size  619

It took  1000  iterations to arrive at the desired result
--- 0.1065676212310791 seconds ---

0.3 2.0

# As observed, we are getting the most optimal values for a and b using a batch size of __619__. Time taken __~0.1067__ secs

In [36]:
time_reduction = (0.2772080898284912 - 0.1065676212310791)/0.2772080898284912

In [37]:
time_reduction *= 100
print(time_reduction)

61.55681412580255


### Q1.e) Does SGD do better or worse in terms of time performance on our data? Is there an optimal minibatch size that works best? Quantify and interpret your findings.

# Inference: 
## With fixed values for parameters like **learn rate = 0.05, and max_iter = 1000**, we can clearly obseve an execution time reduction of **~62%** when we use Stochastic Mini Batch GD. 
## However, it took __1000__ iterations even with the most optimum batch size in Mini Batch Stochastic gradient descent where as it took __706__ in batch version. Therefore, in terms of using CPU, Batch seems to be a better option than Mini Batch.
